### DISCUSSION TITLE
fuse permission problems

### DISCUSSION THREAD

**Discussion by stsp** [libfuse/libfuse#991] [General]
These problems not directly related to
libfuse, but to the kernel, though I tried
to report them in other places in vain.
Maybe someone here can help me.

I have 2 problems with FUSE permissions
that are not present with other FSes.

1. fuse insists on saved-UID to match owner UID.
In fact, fuse_permissible_uidgid() in fs/fuse/dir.c:
https://github.com/torvalds/linux/blob/master/fs/fuse/dir.c#L1421
checks everything but fsuid, whereas other
FSes seem to check fsuid.
Can fuse change that and allow saved-UID
to mismatch? Perhaps by just checking fsuid
instead?

2. My app uses the "file server" which passes
the opened fds to the less-privileged process.
Lower-priv process can't open the files on its
own, but using the pre-opened FDs usually does
the trick.
But this doesn't work with fuse: the passed fd
gives EACCES on eg fstat() (and likely also on
all other syscalls, haven't checked further),
so it seems like the permissions are re-checked,
even though FD is pre-opened.
I suspect this check to be a culprit:
https://github.com/torvalds/linux/blob/master/fs/fuse/dir.c#L1516
It doesn't seem like it checks if FD is pre-opened
or not.

Where should I report these problems?
Tried here:
https://lore.kernel.org/all/9fb28d29-d566-4d96-a491-8f0fbe2e853b@yandex.ru/T/
And here:
https://bugzilla.kernel.org/show_bug.cgi?id=219028
and nothing helps.

**Comment by bsbernd** [comment]
Добрый день @stsp ,

I'm going to convert this into a discussion, as this is not a libfuse issue. Yes, I had seen your message in fsdevel, but I have to point out that neither that list nor here libfuse/github provides 24/7 support.  Firstly this is not payed support and then it is vacation time (I _should_ actually leave today),  но в моем слушае у детей етсть очень высокая температура и я тоже болел...

I have some doubts about fuse_allow_current_process(), it should not disallow if `allow_other` is set and I guess the other process runs in the same namespace. 
It would be helpful, if you could write a small reproducer and then please send that with a reply to linux-fsdevel (feel free to take me in CC. However, as soon as we leave I will be offline for two weeks).

**Reply by stsp** [reply]
Ok, thanks for a reply.
Since the problem came from AppImage
technology, I'll be able to pack a reproducer.
Note however that `allow_other` is not
enabled in my case. AppImage will likely
not want to use `allow_other` for me, and
the whole point is that pre-opened FDs
should work without `allow_other`.

Let me clarify:
- At initial stage, an app is running as an
  owner user, but with altered saved-UID.
  This hits the problem 1.
- Then the file server process if fork()'ed,
  and it has saved-UID aligned to other
  UIDs, so no problems with this one.
- Then it passes the pre-opened FDs to other
  processes, and my point is that this should
  work without `allow_other` because other
  FSes do not re-check perms on a pre-opened
  FDs for most operations (they do for some,
  which I moved to the server side already).

Я тоже простужен, детям скорейшего выздоровления. :)

**Comment by bsbernd** [comment]
Looking at `fuse_getattr` this commit is interesting

```
commit 5157da2ca42cbeca01709e29ce7b797cffed2432
Author: Miklos Szeredi <mszeredi@redhat.com>
Date:   Tue May 19 14:50:37 2020 +0200

    fuse: always allow query of st_dev
    
    Fuse mounts without "allow_other" are off-limits to all non-owners.  Yet it
    makes sense to allow querying st_dev on the root, since this value is
    provided by the kernel, not the userspace filesystem.
    
    Allow statx(2) with a zero request mask to succeed on a fuse mounts for all
    users.
    
    Reported-by: Nikolaus Rath <Nikolaus@rath.org>
    Signed-off-by: Miklos Szeredi <mszeredi@redhat.com>
```

I.e. `Fuse mounts without "allow_other" are off-limits to all non-owners.`

From my point of view, without `allow_other` fuse behaves quite differently than other file systems - the issue you run into is justified.

**Reply by stsp** [reply]
> From my point of view, without allow_other fuse behaves quite differently than other file systems - the issue
> you run into is justified.

It seems `statx()` uses `dirfd` like `fstatat()`,
not an `fd` like `fstat()`. So of course `statx()`
re-checks the permissions, as there is a
`pathname` argument still involved.
All `*at()` funcs re-check permissions.

> We can discuss with Miklos if we can add in the fsid into fuse_permissible_uidgid.

Note that with `allow_other` it is not even
called. So if we conclude `allow_other` is
actually needed (which I hope not), there
would be nothing to discuss with him.

**Comment by bsbernd** [comment]
We can discuss with Miklos if we can add in the fsid into fuse_permissible_uidgid.

**Comment by bsbernd** [comment]
Спасибо! Тебе тоже желаю скорейшего выздоровления!

**Comment by stsp** [comment]
I've finally managed to create a
buildroot setup with a test-case.
Sorry, can't post it to ML, as its too
difficult to re-create anyway.

My findings are:
- The guess about fuse_permissible_uidgid() was correct.
- The guess about second problem was not correct.

The check that prevents the pre-opened FD
operations appeared to be this one:
https://github.com/torvalds/linux/blob/master/fs/fuse/dir.c#L2112
So if I remove it and adjust fuse_permissible_uidgid()
to use fsuid/fsgid, the test-case passes.
But what is this `fuse_getattr()` all about?
And should I remove the similar check from
`fuse_setattr()` too?
What would be the correct fix?

**Comment by bsbernd** [comment]
You need to discuss this with Miklos. But I don't think you can remove the check for `fuse_allow_current_process` - that is the entire point about `allow_other`. And other file systems simply do not have that option.

**Reply by stsp** [reply]
> But I don't think you can remove the check for fuse_allow_current_process

No, not entirely.
It is used in many places, and I want
to remove it only from `fuse_getattr()`.
But I don't know what `fuse_getattr()`
does, and there is also `fuse_setattr()`
with the same check...
Here's the patch:
```
--- dir.c.old   2024-04-13 14:10:12.000000000 +0300
+++ dir.c       2024-07-30 15:00:17.468546562 +0300
@@ -1412,12 +1412,8 @@
 {
        const struct cred *cred = current_cred();

-       return (uid_eq(cred->euid, fc->user_id) &&
-               uid_eq(cred->suid, fc->user_id) &&
-               uid_eq(cred->uid,  fc->user_id) &&
-               gid_eq(cred->egid, fc->group_id) &&
-               gid_eq(cred->sgid, fc->group_id) &&
-               gid_eq(cred->gid,  fc->group_id));
+       return (uid_eq(cred->fsuid, fc->user_id) &&
+               gid_eq(cred->fsgid, fc->group_id));
 }

 /*
@@ -2076,24 +2072,10 @@
                        u32 request_mask, unsigned int flags)
 {
        struct inode *inode = d_inode(path->dentry);
-       struct fuse_conn *fc = get_fuse_conn(inode);

        if (fuse_is_bad(inode))
                return -EIO;

-       if (!fuse_allow_current_process(fc)) {
-               if (!request_mask) {
-                       /*
-                        * If user explicitly requested *nothing* then don't
-                        * error out, but return st_dev only.
-                        */
-                       stat->result_mask = 0;
-                       stat->dev = inode->i_sb->s_dev;
-                       return 0;
-               }
-               return -EACCES;
-       }
-
        return fuse_update_get_attr(inode, NULL, stat, request_mask, flags);
 }
 
```

**Reply by bsbernd** [reply]
I think you want to use `git diff` or `diff -p` to include function names in diffs.  Why would you *replace* euid/egid and uid/gid and not just add in fsuid? 

getattr -> stat
setattr -> update stat and time stamps

Your patch basically allows other processes that had been forbidden before - why would that be correct?

**Reply by stsp** [reply]
> Why would you replace euid/egid and uid/gid and not just add in fsuid?

Because my problem is that I get
`EACCES` when either of UIDs do not
match, even saved-uid. If I only add
fsuid check and remove nothing, my
problem is not solved. I think other
FSes check only fsuid, is this wrong?

> getattr -> stat

Ok, thanks. My test-case does exactly
`fstat()`. :) So obviously that removal is
not fixing the entire thing.
How can I distinguish `stat()` and `fstat()`?
I only need the pre-opened FD to bypass
the check, which is `fstat()` but not `stat()`.

> Your patch basically allows other processes that had been forbidden before - why would that be correct?

It can't be correct, I don't even understand
what I remove. :) I just wrote a stupid test-case
with `fstat()` and made sure it passes.
So you explained the "attr" part to me - thanks!
Next step:
How to distinguish fd-based ops with path-based
ops?

**Reply by bsbernd** [reply]
You are on kernel level - fuse-kernel has no clue about the path. "fd-based ops with path-based ops" do not exist in that form.

**Reply by stsp** [reply]
So basically there is no possibility
to relax the check for pre-opened fds?

**Comment by bsbernd** [comment]
Not like this. The vfs knows at the beginning of the syscalls if this is fstat or stat*, but doesn't pass through this information - would require a larger code updates and convincing people why that is needed. Easier is to add in allow_other

**Reply by stsp** [reply]
> Easier is to add in allow_other

I am all for adding allow_other, but
it requires root! You need to edit
`/etc/fuse.conf` uncommenting
`user_allow_other`.

**Reply by stsp** [reply]
> The vfs knows at the beginning of the syscalls if this is fstat or stat*, but doesn't pass through this information

Doesn't VFS do the permission checks
on its own already? Maybe the fuse
checks are redundant?

**Reply by stsp** [reply]
Or is it like this: by default FUSE
does more restrictive checks, but
with "allow_other" it relies on the
VFS checks?

**Comment by bsbernd** [comment]
Yes exactly, see this comment
```
/*
 * Calling into a user-controlled filesystem gives the filesystem
 * daemon ptrace-like capabilities over the current process.  This
 * means, that the filesystem daemon is able to record the exact
 * filesystem operations performed, and can also control the behavior
 * of the requester process in otherwise impossible ways.  For example
 * it can delay the operation for arbitrary length of time allowing
 * DoS against the requester.
 *
 * For this reason only those processes can call into the filesystem,
 * for which the owner of the mount has ptrace privilege.  This
 * excludes processes started by other users, suid or sgid processes.
 */
 bool fuse_allow_current_process(struct fuse_conn *fc)
```

**Reply by stsp** [reply]
Its very difficult to parse this comment
properly. Am I right that "call into the
filesystem" means client, and "owner of
the mount" means FS server?

So if this all is done to spare the client
from being attacked by an FS server,
then it seems logical to allow the client
to "accept the risk" and say "please allow
me to see this mount, I trust the FS
server to not DoS me and I really need
this mount".
That looks much safer than making
the mount system-wide visible with
"allow_others". What do you think?

**Comment by stsp** [comment]
I hit yet another problem with permissions
and filled this ticket:
https://github.com/containers/fuse-overlayfs/issues/428
It has the test-case within.
Maybe you can help with that one too. :)

**Comment by bsbernd** [comment]
@stsp  sorry, I had missed this. I would need to read what is the issue is - I'm afraid I don't have the time.

**Comment by stsp** [comment]
/etc/fuse.conf is a user-space issue btw.
I wonder if the different default is acceptable.

**Comment by bsbernd** [comment]
Typically that is a distribution and actually admin decision - installing the libfuse package should pop up a question. These are security concerns - we need to be careful here. Though fully I understand your issue.

**Reply by stsp** [reply]
But isn't this option unprivileged in the kernel?
Or have I misunderstood Miklos's reply and
"allow_other" actually needs caps?

**Comment by bsbernd** [comment]
I don't think it needs caps, but let's say an admin allows users to spawn their own fuse processes, but does not want that other users can access these mounts - admin might not notice this change and create a security issue. Problem here are multi user environments...

**Reply by stsp** [reply]
But if it doesn't require caps, then
who prevents an attacker from using
his own patched fusermount?

**Comment by bsbernd** [comment]
fusermount needs to have the s-bit for that. 
```
bernd@e7270 ~>ll /usr/bin/fusermount3
-rwsr-xr-x 1 root root 39K Apr  8 17:57 /usr/bin/fusermount3
```

**Reply by stsp** [reply]
Do you mean, fuse mount operation requires
caps anyway, just that "allow_other" doesn't
add any?

**Comment by bsbernd** [comment]
And sorry, I had mis-understood your question - entire fusermount needs to privileged/have CAP_SYS_ADMIN.

**Reply by stsp** [reply]
But in the nearby thread (which you said
you haven't time to look into), I patched
fuse-overlayfs and provided the patched
version with my project. It works w/o any
caps. So I was under the deep impression
that fuse can be mounted w/o any caps,
but was that wrong?

**Comment by bsbernd** [comment]
See `man 2 mount`


```
       Appropriate privilege (Linux: the CAP_SYS_ADMIN capability) is required to mount filesystems.
```

**Reply by stsp** [reply]
But could you please explain the following:
```
$ ls -l `which fusermount3`
-rwsr-xr-x 1 root root 39296 апр  8 18:57 /usr/bin/fusermount3
$ ls -l `which fuse-overlayfs`
-rwxr-xr-x 1 root root 101096 ноя 27  2023 /usr/bin/fuse-overlayfs
```

fusermount3 has `s` but fuse-overlayfs doesn't.
And I really already patched fuse-overlayfs for
my project. Why are they different?

**Comment by bsbernd** [comment]
When you run fuse-overlayfs as root or CAP_SYS_ADMIN, plain mount() will succeed and fusermount is not called. When mount() fails, it falls back to fusermount. See `fuse_mount_sys` and the caller `fuse_kern_mount`in `lib/mount.c`

**Reply by stsp** [reply]
Oh. :(
Thanks for explaining.
